themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,205 @@
+# previews. This field is required or the build will fail the linter test.
+# Max length is 160 characters.
+meta_desc: Pulumi is often debated to be imperative or declarative.
+  What if I tell you that Pulumi is imperative, declarative and imperative?
+
+# The meta_image appears in social-media previews and on the blog home page.
+# A placeholder image representing the recommended format, dimensions and aspect
+# ratio has been provided for you.
+meta_image: meta.png
+
+# At least one author is required. The values in this list correspond with the `id`
+# properties of the team member files at /data/team/team. Create a file for yourself
+# if you don't already have one.
+authors:
+    - ringo-de-smet
+
+# At least one tag is required. Lowercase, hyphen-delimited is recommended.
+tags:
+    - cloud-engineering
+    - infrastructure-as-code
+
+# See the blogging docs at https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md.
+# for additional details, and please remove these comments before submitting for review.
+---
+
+<!--
+Summary:
+- show single binary setup of Terraform: HCL processing & Engine
+- show double binary setup of Pulumi: Language Host & Engine
+- indicate that the chosen language (imperative or not) doesn't have any implications of the engine remaining declarative
+- we chose a gRPC protocol, rather than an in-process call to request creation of resources to the engine
+- the provisioning magic happens in the engine & the providers
+- now comes the nifty part
+- with this separation, Pulumi brings cloud engineering closer to developers
+- we allow for components (Pulumi Packages) to be written in any language given a component is a resource by itself.
+- we allow for policies to be written with the same ease as abstractions.
+- we mix it all in a single setup
+
+Sketch the setup of the article as the description of the PR for reviewers
+-->
+
+On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
+It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?
+
+<!--more-->
+
+When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:
+
+> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
+> It’s not used for talking to the cloud provider API.
+
+The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
+into our product.
+
+{{< tweet id="1553431913691430912" user="adamhjk" >}}
+
+Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.
+Let me use the Pulumi architecture to highlight why I mention *imperative* twice.
```

Comment:
Would it be a good idea to define `imperative` and `declaritve`?  Something like this:

`Both terms refer to how the user provides direction to the automation platform. With an imperative tool, you define the steps to execute in order to reach the desired solution. With a declarative tool, you define the desired state of the final solution, and the automation platform determines how to achieve that state.`

I got the above from the following [link](https://www.linode.com/blog/devops/declarative-vs-imperative-in-iac/)